### PR TITLE
UCP/WIRUEP/GTEST/UCT: Fix not selecting RNDV/RMA/AMO lanes during CM init phase

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

1. Fix not selecting RNDV/RMA/AMO lanes during CM init phase.
2. Add gtest to check RNDV lanes selection.
3. Add support for `UCT_MD_FLAG_INVALIDATE` flags in TCP.

## Why ?

1. To use RNDV GET/PUT Zcopy schemes for EP created when `UCX_CM_USE_ALL_DEVICES=n` is set.
2. To reproduce the problem.
3. To pass gtest for TCP.

## How ?

1. Added support `UCP_EP_INIT_CREATE_AM_LANE_ONLY` EP init flags; use this flag when creating CM client initial config to fit CM data length limit.
2. If PUT/GET Zcopy capabilities are supported by EP, check that RNDV lanes are created.
3. To pass the new gtest for TCP.

### CODE REVIEW

File: src/ucp/core/ucp_request.c [hunk 1]
```diff
@@ -308,7 +308,10 @@ static ucp_md_map_t ucp_request_get_invalidation_map(ucp_request_t *req)
         lane      = key->rma_bw_lanes[i];
         cap_flags = ucp_ep_get_iface_attr(ep, lane)->cap.flags;
 
-        if (cap_flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) {
+        if ((cap_flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) &&
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c:311]: 1. seems unrelated fix?
2. maybe use ucp_ep_is_lane_p2p?

File: src/uct/tcp/tcp_md.c [hunk 2]
```diff
@@ -16,6 +16,7 @@ static ucs_status_t uct_tcp_md_query(uct_md_h md, uct_md_attr_t *attr)
 {
     /* Dummy memory registration provided. No real memory handling exists */
     attr->cap.flags               = UCT_MD_FLAG_REG |
+                                    UCT_MD_FLAG_INVALIDATE |
```
Comment (yosefe) [inline src/uct/tcp/tcp_md.c:19]: but TCP does not really support invalidate?

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -396,9 +397,20 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
...
+        local_md_flags = criteria->local_md_flags;
+        if (!(iface_attr->cap.flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) ||
+            /* If the transport has both CONNECT_TO_IFACE and CONNECT_TO_EP, and
+             * connecting via CM, UCP prefers CONNECT_TO_EP */
+            ((iface_attr->cap.flags & UCT_IFACE_FLAG_CONNECT_TO_EP) &&
+             (ucp_ep_init_flags_has_cm(select_params->ep_init_flags)))) {
+            /* Remove INVALIDATE md flag in case of transport uses CONNECT_TO_EP
+             * connection mode */
+            local_md_flags &= ~UCT_MD_FLAG_INVALIDATE;
+        }
```
Comment (yosefe) [inline src/ucp/wireup/select.c:409]: why need to remove flag? better not to add in first place if not needed..

File: src/ucp/core/ucp_request.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:311]: it is related to get the new gtest passing for TCP

File: src/uct/tcp/tcp_md.c (hunk 2 above)
Comment (dmitrygx) [reply src/uct/tcp/tcp_md.c:19]: TCP doesn't support rkey actually, it is just no-op.
so, we could do invalidate for TCP to support RNDV for TCP

Comment (yosefe) [reply src/uct/tcp/tcp_md.c:19]: IMO, the opposite: what prevents from sending PUT request to a released buffer?

File: test/gtest/common/test_helpers.cc [hunk 4]
```diff
@@ -607,7 +607,7 @@ sock_addr_storage::sock_addr_storage(const ucs_sock_addr_t &ucs_sock_addr,
                       is_rdmacm_netdev);
     }
 }
-
+
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply test/gtest/common/test_helpers.cc]: removed

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/wireup/select.c]: will be done in the separate PR

File: src/ucp/wireup/wireup_cm.c [hunk 5]
```diff
@@ -424,9 +428,18 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
+            ep, &tl_bitmap, dev_index,
+            (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) ||
+            (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE),
+            &priv_data, &priv_data_length);
     if (status == UCS_ERR_BUFFER_TOO_SMALL) {
+        if (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) {
+            /* Create endpoint configuration with AM lane only to fit CM private
+             * data length limit */
+            ep_init_flags = UCP_EP_INIT_CREATE_AM_LANE_ONLY;
+            goto initial_config_get;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: done

File: src/uct/tcp/tcp_md.c (hunk 2 above)
Comment (dmitrygx) [reply src/uct/tcp/tcp_md.c]: yes, need to add support in TCP in the near future

File: src/ucp/core/ucp_request.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.c]: will be done in the separate PR when this PR gets merged

**Comment by dmitrygx** [comment]
@yosefe @brminich could you review pls?

File: src/ucp/wireup/wireup_cm.c [hunk 6]
```diff
@@ -424,9 +427,20 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
     /* Replay pending requests from the tmp_pending_queue */
     ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
 
-    status = ucp_cm_ep_priv_data_pack(ep, &tl_bitmap, dev_index, &priv_data,
-                                      &priv_data_length);
+
+    status = ucp_cm_ep_priv_data_pack(
+            ep, &tl_bitmap, dev_index,
+            (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) ||
+            (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE),
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:434]: use helper var

File: src/ucp/wireup/wireup_cm.c [hunk 7]
```diff
@@ -1199,7 +1213,7 @@ ucp_ep_server_init_priv_data(ucp_ep_h ep,  const char *dev_name,
     ucp_tl_bitmap_validate(&tl_bitmap, &ctx_tl_bitmap);
 
     dev_index = ucp_cm_tl_bitmap_get_dev_idx(worker->context, &tl_bitmap);
-    status    = ucp_cm_ep_priv_data_pack(ep, &tl_bitmap, dev_index,
+    status    = ucp_cm_ep_priv_data_pack(ep, &tl_bitmap, dev_index, 0,
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:1216]: before this PR, can_fallback was equivalent to ucp_cm_client_get_next_cm_idx(ep)

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 8]
```diff
@@ -1221,6 +1222,57 @@ UCS_TEST_P(test_ucp_sockaddr_cm_private_data,
 UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_sockaddr_cm_private_data, all, "all")
 
 
+class test_ucp_sockaddr_check_lanes : public test_ucp_sockaddr {
+    unsigned rndv_lanes_num(ucp_ep_h ep)
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:1226]:
```
private:
   bool has_rma_bw_lanes()  { 
      return ucp_ep_config(ep)->key.rma_bw_lanes[0] != UCP_NULL_LANE; 
   }
```

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 9]
```diff
@@ -1221,6 +1222,57 @@ UCS_TEST_P(test_ucp_sockaddr_cm_private_data,
...
+                (ucp_ep_get_iface_attr(ep, lane_idx)->cap.flags &
+                         (UCT_IFACE_FLAG_GET_ZCOPY |
+                          UCT_IFACE_FLAG_PUT_ZCOPY)) &&
+                /* RNDV lanes are not selected for transports which don't
+                 * support UCT_MD_FLAG_INVALIDATE capability on MD.
+                 * TODO: fix by allowing CONNECT_TO_EP transports being selected
+                 *       without UCT_MD_FLAG_INVALIDATE capability */
+                (ucp_ep_md_attr(ep, lane_idx)->cap.flags &
+                         UCT_MD_FLAG_INVALIDATE)) {
+                EXPECT_GT(rndv_lanes_num(ep), 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:1256]: EXPECT_NE(UCP_NULL_LANE, ucp_ep_config(ep)->key.rma_bw_lanes[0]) << "RNDV lanes should be selected"

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 10]
```diff
@@ -1221,6 +1222,57 @@ UCS_TEST_P(test_ucp_sockaddr_cm_private_data,
...
+        for (ucp_lane_index_t lane_idx = 0;
+             lane_idx < ucp_ep_num_lanes(ep); ++lane_idx) {
+            if ((lane_idx != ucp_ep_get_cm_lane(ep)) &&
+                (ucp_ep_get_iface_attr(ep, lane_idx)->cap.flags &
+                         (UCT_IFACE_FLAG_GET_ZCOPY |
+                          UCT_IFACE_FLAG_PUT_ZCOPY)) &&
+                /* RNDV lanes are not selected for transports which don't
+                 * support UCT_MD_FLAG_INVALIDATE capability on MD.
+                 * TODO: fix by allowing CONNECT_TO_EP transports being selected
+                 *       without UCT_MD_FLAG_INVALIDATE capability */
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:1253]: /* RNDV lanes should be selected if transport supports get/put_zcopy and invalidate */

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 9 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_sockaddr.cc:1256]: done, removed `has_rma_bw_lanes()` as not needed in this case

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]
